plugins/catalog-backend/src/processing/TaskPipeline.ts

```diff
@@ -72,50 +72,112 @@ export function startTaskPipeline<T>(options: Options<T>) {
+              .finally(() => {
+                state.inFlightCount -= 1;
+                barrier.release();
+              });
+          }
+        }
+      }
 
-  async function maybeLoadMore() {
-    if (stopped || loading || inFlightCount > lowWatermark) {
-      return;
+      await barrier.wait();
     }
+  }
 
-    // Once we hit the low watermark we load in enough items to reach the high watermark
-    loading = true;
-    const loadCount = highWatermark - inFlightCount;
-    const loadedItems = await loadTasks(loadCount);
-    loading = false;
-
-    // We might not reach the high watermark here, in case there weren't enough items to load
-    inFlightCount += loadedItems.length;
-    loadedItems.forEach(item => {
-      processTask(item).finally(() => {
-        if (stopped) {
-          return;
-        }
+  pipelineLoop().catch(error => {
+    // This should be impossible, but if it did happen, it would signal a
+    // programming error inside the loop (errors should definitely be caught
+    // inside of it). Let's rethrow with more information, and let it be caught
+    // by the process' uncaught exception handler, which will log the occurrence
+    // at a high level.
+    throw new Error(`Unexpected error in processing pipeline loop`, error);
+  });
+
+  return () => {
+    abortController.abort();
+    barrier.destroy();
+  };
+}
 
-        // For each item we complete we check if it's time to load more
-        inFlightCount -= 1;
-        maybeLoadMore();
-      });
+/**
+ * Creates a barrier with a timeout, that can be awaited or prematurely
+ * released either manually or by an abort signal.
+ */
+export function createBarrier(options: {
+  waitTimeoutMillis: number;
+  signal: AbortSignal;
+}): {
+  wait: () => Promise<void>;
+  release: () => void;
+  destroy: () => void;
+} {
+  const { waitTimeoutMillis, signal } = options;
+  const resolvers = new Set<() => void>();
+
+  function wait() {
+    if (signal.aborted || !(waitTimeoutMillis > 0)) {
+      return Promise.resolve();
+    }
+
+    return new Promise<void>(resolve => {
+      const timeoutHandle = setTimeout(done, waitTimeoutMillis);
+
+      function done() {
+        resolvers.delete(done);
+        clearTimeout(timeoutHandle);
+        resolve();
+      }
+
+      resolvers.add(done);
     });
+  }
 
-    // We might have processed some tasks while we where loading, so check if we can load more
-    if (loadedItems.length > 1) {
-      maybeLoadMore();
+  function release() {
+    const resolversToCall = new Set(resolvers);
+    resolvers.clear();
+    for (const resolver of resolversToCall) {
+      resolver();
     }
   }
 
-  // This interval makes sure that we load in new items if the loop runs
-  // dry because of the lack of available tasks. As long as there are
-  // enough items to process this will be a noop.
-  const intervalId = setInterval(() => {
-    maybeLoadMore();
-  }, pollingIntervalMs);
+  signal.addEventListener('abort', release);
```

Comment:
hodl, just removing it in the callback would be enough right? As in

```ts
  function release() {
    const resolversToCall = new Set(resolvers);
    resolvers.clear();
    for (const resolver of resolversToCall) {
      resolver();
    }
    signal.removeEventListener('abort', release)
  }
```

Reply:
No, release is called over and over again. It'd render this subscription immediately useless after the first task completes. Note that this addition is made only exactly once.

Reply:
Well with a different function for the signal abort callback

```ts
function onAbort() {
  release()
  signal.removeEventListener('abort', onAbort)
}
signal.addEventListener('abort', onAbort);
```

Reply:
Mm. I wonder how much we need to worry though. It uses a weak map internally afaik so it should not leak